sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Processor/EventProcessorClientLiveTests.cs

```diff
@@ -31,5 +36,831 @@ namespace Azure.Messaging.EventHubs.Tests
+        /// <param name="batchEvents">A dictionary to which the events included in the batches should be tracked.</param>
+        /// <param name="cancellationToken">The token used to signal a cancellation request.</param>
+        ///
+        /// <returns>The set of batches needed to contain the entire set of <paramref name="events"/>.</returns>
+        ///
+        /// <remarks>
+        ///   Callers are assumed to be responsible for taking ownership of the lifespan of the returned batches, including
+        ///   their disposal.
+        ///
+        ///   This method is intended for use within the test suite only; it is not hardened for general purpose use.
+        /// </remarks>
+        ///
+        private async Task<IEnumerable<EventDataBatch>> BuildBatchesAsync(IEnumerable<EventData> events,
+                                                                          EventHubProducerClient producer,
+                                                                          CancellationToken cancellationToken)
+        {
+            EventData eventData;
+
+            var queuedEvents = new Queue<EventData>(events);
+            var batches = new List<EventDataBatch>();
+            var currentBatch = default(EventDataBatch);
+
+            while (queuedEvents.Count > 0)
+            {
+                currentBatch ??= (await producer.CreateBatchAsync(cancellationToken).ConfigureAwait(false));
+                eventData = queuedEvents.Peek();
+
+                if (!currentBatch.TryAdd(eventData))
+                {
+                    if (currentBatch.Count == 0)
+                    {
+                        throw new InvalidOperationException("There was an event too large to fit into a batch.");
+                    }
+
+                    batches.Add(currentBatch);
+                    currentBatch = default;
+                }
+                else
+                {
+                   queuedEvents.Dequeue();
+                }
+            }
+
+            if ((currentBatch != default) && (currentBatch.Count > 0))
+            {
+                batches.Add(currentBatch);
+            }
+
+            return batches;
+        }
+
+        /// <summary>
+        ///   Attempts to add an item to the <paramref name="dictionary" /> if it does not
+        ///   already exist.
+        /// </summary>
+        ///
+        /// <typeparam name="TKey">The type of the key.</typeparam>
+        /// <typeparam name="TValue">The type of the value.</typeparam>
+        ///
+        /// <param name="dictionary">The dictionary to attempt to add the item to.</param>
+        /// <param name="key">The key to assign the item.</param>
+        /// <param name="value">The value of the item.</param>
+        ///
+        /// <returns><c>true</c> if the item was added; otherwise, <c>false</c>.</returns>
+        ///
+        private static bool TryAdd<TKey, TValue>(Dictionary<TKey, TValue> dictionary,
```

Comment:
Why not use a `ConcurrentDictionary` which already implements `TryAdd`? I believe this implementation could throw on `dictionary.Add` in a concurrent scenario.

Reply:
I probably should have just used a concurrent dictionary...  especially now that I kicked to more than a single partition.   I was originally creating just one, so there was never a chance of concurrent calls.   I'll address this in a quick follow-up change set.